File: examples/cpp/common/rdp_connection.cpp
// Copyright (c) 2026 solar@heliacal.net
// SPDX-License-Identifier: MIT

#include "rdp_connection.h"

RDPMessage::RDPMessage()
    : m_message(nullptr)
{
}

RDPMessage::~RDPMessage()
{
    Reset();
}

const std::uint8_t *RDPMessage::Data() const
{
    return m_message != nullptr ? static_cast<const std::uint8_t *>(rdplib_message_data(m_message)) : nullptr;
}

std::uint32_t RDPMessage::Size() const
{
    return m_message != nullptr ? rdplib_message_size(m_message) : 0;
}

std::uint8_t RDPMessage::StreamNumber() const
{
    if (m_message == nullptr || (rdplib_message_flags(m_message) & RDP_FLAG_SEQUENCED) == 0)
        return 0;

    return rdplib_message_stream(m_message);
}

std::uint16_t RDPMessage::Flags() const
{
    return m_message != nullptr ? rdplib_message_flags(m_message) : 0;
}

void RDPMessage::Reset(rdplib_message_t *message)
{
    if (m_message != nullptr)
        rdplib_message_release(m_message);

    m_message = message;
}

RDPConnection::RDPConnection(rdplib_connection_t *connection)
    : m_connection(connection),
      m_terminal_result(NoData),
      m_disconnect_reason(0)
{
}

RDPConnection::~RDPConnection()
{
    Close(0);
}

int RDPConnection::EnableKeepalive()
{
    return m_connection != nullptr ? rdplib_connection_enable_keepalive(m_connection) : RDPLIB_ERROR_NOT_USABLE;
}

int RDPConnection::SetDataRate(std::uint32_t bytes_per_second)
{
    return m_connection != nullptr ? rdplib_connection_set_data_rate(m_connection, bytes_per_second) : RDPLIB_ERROR_NOT_USABLE;
}

int RDPConnection::SetSendBufferSize(std::uint32_t bytes)
{
    return m_connection != nullptr ? rdplib_connection_set_send_buffer_size(m_connection, bytes) : RDPLIB_ERROR_NOT_USABLE;
}

int RDPConnection::Send(const void *data, std::uint32_t bytes, std::uint32_t stream, std::uint32_t flags)
{
    if (m_connection == nullptr || m_terminal_result != NoData)
        return RDPLIB_ERROR_NOT_USABLE;

    int result = rdplib_connection_send(m_connection, data, bytes, stream, flags);
    bool reliable = (flags & RDPLIB_SEND_RELIABLE) != 0;
    bool connection_lost = result == RDPLIB_CONNECTION_SEND_HISTORY_FULL ||
        (result == RDPLIB_CONNECTION_SEND_BUFFER_FULL && reliable) ||
        result == RDPLIB_CONNECTION_SEND_NOT_CONNECTED ||
        result == RDPLIB_CONNECTION_SEND_FIN_SENT ||
        result == RDPLIB_CONNECTION_SEND_PEER_STOPPED;

    if (connection_lost)
    {
        m_terminal_result = ConnectionLost;
        m_disconnect_reason = RDPLIB_DISCONNECT_REASON_SEND_ERROR;
    }

    return result;
}

RDPConnection::ReceiveResult RDPConnection::Receive(RDPMessage *message, std::uint32_t *disconnect_reason)
{
    if (message == nullptr)
        return NoData;

    message->Reset();
    if (disconnect_reason != nullptr)
        *disconnect_reason = 0;

    if (m_terminal_result != NoData)
    {
        if (disconnect_reason != nullptr && m_terminal_result == ConnectionLost)
            *disconnect_reason = m_disconnect_reason;

        return m_terminal_result;
    }

    if (m_connection == nullptr)
        return NoData;

    rdplib_message_t *received_message = rdplib_connection_pop_message(m_connection);
    if (received_message == nullptr)
        return NoData;

    if (rdplib_message_is_disconnect(received_message))
    {
        rdplib_disconnect_info_t information{};
        if (rdplib_connection_get_disconnect_info(m_connection, &information) == RDPLIB_OK)
            m_disconnect_reason = information.reason;

        rdplib_message_release(received_message);
        m_terminal_result = ConnectionLost;
        if (disconnect_reason != nullptr)
            *disconnect_reason = m_disconnect_reason;

        return m_terminal_result;
    }

    bool has_fin = rdplib_message_has_fin(received_message) != 0;
    std::uint32_t bytes = rdplib_message_size(received_message);
    if (has_fin)
        m_terminal_result = PeerClosed;

    if (has_fin && bytes == 0)
    {
        rdplib_message_release(received_message);
        return PeerClosed;
    }

    message->Reset(received_message);
    return MessageReceived;
}

void RDPConnection::DiscardMessages()
{
    if (m_connection == nullptr)
        return;

    rdplib_message_t *message;
    while ((message = rdplib_connection_pop_message(m_connection)) != nullptr)
        rdplib_message_release(message);
}

void RDPConnection::Close(std::uint32_t linger_timeout_ms)
{
    if (m_connection == nullptr)
        return;

    DiscardMessages();

    rdplib_connection_t *connection = m_connection;
    m_connection = nullptr;

    (void)rdplib_connection_begin_close(connection, linger_timeout_ms);
    rdplib_connection_release(connection);
}

int RDPConnection::GetRemoteAddress(std::uint8_t address[4], std::uint16_t &port) const
{
    return m_connection != nullptr ? rdplib_connection_get_remote_ipv4(m_connection, address, &port) : RDPLIB_ERROR_NOT_USABLE;
}

File: examples/dotnet/common/RdpConnection.cs
// Copyright (c) 2026 solar
// SPDX-License-Identifier: MIT

using System;
using System.Runtime.InteropServices;

namespace RdplibExample.Common
{
    internal sealed class RdpMessage
    {
        internal byte[] Data { get; private set; }
        internal byte StreamNumber { get; private set; }
        internal ushort Flags { get; private set; }

        internal RdpMessage()
        {
            Data = Array.Empty<byte>();
        }

        internal void Reset(byte[] data = null, byte streamNumber = 0, ushort flags = 0)
        {
            Data = data ?? Array.Empty<byte>();
            StreamNumber = streamNumber;
            Flags = flags;
        }
    }

    // Owns one application connection handle. Its endpoint and runtime must
    // outlive it.
    internal sealed class RdpConnection : IDisposable
    {
        internal enum ReceiveResult
        {
            NoData,
            MessageReceived,
            PeerClosed,
            ConnectionLost
        }

        private IntPtr _connection;
        private ReceiveResult _terminalResult;
        private uint _disconnectReason;

        internal RdpConnection(IntPtr connection)
        {
            _connection = connection;
            _terminalResult = ReceiveResult.NoData;
        }

        internal int EnableKeepalive()
        {
            return _connection != IntPtr.Zero
                ? RdplibNative.rdplib_connection_enable_keepalive(_connection)
                : RdplibNative.ErrorNotUsable;
        }

        internal int SetDataRate(uint bytesPerSecond)
        {
            return _connection != IntPtr.Zero
                ? RdplibNative.rdplib_connection_set_data_rate(_connection, bytesPerSecond)
                : RdplibNative.ErrorNotUsable;
        }

        internal int SetSendBufferSize(uint bytes)
        {
            return _connection != IntPtr.Zero
                ? RdplibNative.rdplib_connection_set_send_buffer_size(_connection, bytes)
                : RdplibNative.ErrorNotUsable;
        }

        // rdplib copies the borrowed data before returning.
        internal int Send(byte[] data, uint stream, uint flags)
        {
            if (_connection == IntPtr.Zero || _terminalResult != ReceiveResult.NoData || data == null)
                return RdplibNative.ErrorNotUsable;

            int result = RdplibNative.rdplib_connection_send(
                _connection,
                data,
                checked((uint)data.Length),
                stream,
                flags
            );

            bool reliable = (flags & RdplibNative.SendReliable) != 0;
            bool connectionLost = result == RdplibNative.ConnectionSendHistoryFull ||
                (result == RdplibNative.ConnectionSendBufferFull && reliable) ||
                result == RdplibNative.ConnectionSendNotConnected ||
                result == RdplibNative.ConnectionSendFinSent ||
                result == RdplibNative.ConnectionSendPeerStopped;

            if (connectionLost)
            {
                _terminalResult = ReceiveResult.ConnectionLost;
                _disconnectReason = RdplibNative.DisconnectReasonSendError;
            }

            return result;
        }

        // Native message data is copied before the native message is released.
        internal ReceiveResult Receive(RdpMessage message, out uint disconnectReason)
        {
            disconnectReason = 0;
            if (message == null)
                return ReceiveResult.NoData;

            message.Reset();
            if (_terminalResult != ReceiveResult.NoData)
            {
                if (_terminalResult == ReceiveResult.ConnectionLost)
                    disconnectReason = _disconnectReason;

                return _terminalResult;
            }

            if (_connection == IntPtr.Zero)
                return ReceiveResult.NoData;

            IntPtr nativeMessage = RdplibNative.rdplib_connection_pop_message(_connection);
            if (nativeMessage == IntPtr.Zero)
                return ReceiveResult.NoData;

            try
            {
                if (RdplibNative.rdplib_message_is_disconnect(nativeMessage) != 0)
                {
                    RdplibNative.DisconnectInfo information;
                    if (RdplibNative.rdplib_connection_get_disconnect_info(_connection, out information) == RdplibNative.Ok)
                        _disconnectReason = information.Reason;

                    _terminalResult = ReceiveResult.ConnectionLost;
                    disconnectReason = _disconnectReason;
                    return _terminalResult;
                }

                bool hasFin = RdplibNative.rdplib_message_has_fin(nativeMessage) != 0;
                uint byteCount = RdplibNative.rdplib_message_size(nativeMessage);
                if (hasFin)
                    _terminalResult = ReceiveResult.PeerClosed;

                if (hasFin && byteCount == 0)
                    return ReceiveResult.PeerClosed;

                int length = checked((int)byteCount);
                byte[] data = new byte[length];
                if (length != 0)
                    Marshal.Copy(RdplibNative.rdplib_message_data(nativeMessage), data, 0, length);

                ushort flags = RdplibNative.rdplib_message_flags(nativeMessage);
                byte streamNumber = (flags & RdplibNative.FlagSequenced) != 0
                    ? RdplibNative.rdplib_message_stream(nativeMessage)
                    : (byte)0;

                message.Reset(data, streamNumber, flags);
                return ReceiveResult.MessageReceived;
            }
            finally
            {
                RdplibNative.rdplib_message_release(nativeMessage);
            }
        }

        // Releases the application handle. The endpoint retains a lingering
        // transport connection until it finishes or its deadline expires.
        internal void Close(uint lingerTimeoutMs)
        {
            if (_connection == IntPtr.Zero)
                return;

            DiscardMessages();

            IntPtr connection = _connection;
            _connection = IntPtr.Zero;
            RdplibNative.rdplib_connection_begin_close(connection, lingerTimeoutMs);
            RdplibNative.rdplib_connection_release(connection);
        }

        internal int GetRemoteAddress(byte[] address, out ushort port)
        {
            port = 0;
            if (_connection == IntPtr.Zero || address == null || address.Length != 4)
                return RdplibNative.ErrorNotUsable;

            return RdplibNative.rdplib_connection_get_remote_ipv4(_connection, address, out port);
        }

        private void DiscardMessages()
        {
            if (_connection == IntPtr.Zero)
                return;

            IntPtr message;
            while ((message = RdplibNative.rdplib_connection_pop_message(_connection)) != IntPtr.Zero)
                RdplibNative.rdplib_message_release(message);
        }

        public void Dispose()
        {
            Close(0);
        }
    }
}
